bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -132,45 +132,51 @@ void ClientsManager::RequestsInfo::logAllPendingRequestsExceedingThreshold(const
 
 /*************************** Class ClientsManager::RepliesInfo ***************************/
 
-void ClientsManager::RepliesInfo::deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient) {
-  Time earliestTime = MaxTime;
-  ReqId earliestReplyId = 0;
-  if (repliesMap_.size() < maxNumOfReqsPerClient) return;
+void ClientsManager::RepliesInfo::deleteReplyIfNeededSafe(NodeIdType clientId,
+                                                          ReqId reqSeqNum,
+                                                          uint16_t maxNumOfReqsPerClient,
+                                                          uint16_t reqOffset) {
+  ReqId oldReplyId = 0;
   if (repliesMap_.size() > maxNumOfReqsPerClient)
     LOG_FATAL(CL_MNGR,
               "More than maxNumOfReqsPerClient_ items in repliesInfo"
                   << KVLOG(repliesMap_.size(), maxNumOfReqsPerClient, clientId));
+  const lock_guard<mutex> lock(repliesMapMutex_);
```

Comment:
Please return the lock back where it was before. I see I forgot to add an explanation for the 'strange' locking model here. The thread safety here is between the PreProcessor thread, which is read-only, and the ReplicaImp main thread, which performs either read or write at a time. So, all write operations should be guarded as well as PrePropessor calls hasReply and isClientRequestInProcess. All other operations are safe. Please add the explaining comment to avoid misunderstanding.

Reply:
Thanks for the explanation.